File: packages/sdk-py/pylva/errors/strict_provider.py
"""Errors raised before a controlled provider request can be dispatched."""

from __future__ import annotations

from typing import Literal

PYLVA_STRICT_PROVIDER_UNSUPPORTED_CODE = "strict_provider_unsupported"

StrictProviderReason = Literal[
    "invalid_client",
    "provider_retries_enabled",
    "unsupported_request_shape",
    "usage_bound_required",
    "unsupported_pricing_feature",
]


class PylvaStrictProviderError(TypeError):
    """A request cannot be bounded and priced by the strict provider surface.

    The exception deliberately contains only a provider and a stable reason. It
    never reflects prompt content, tool arguments, or a provider response body.
    """

    code = PYLVA_STRICT_PROVIDER_UNSUPPORTED_CODE

    def __init__(self, provider: Literal["openai", "anthropic"], reason: StrictProviderReason):
        self.provider = provider
        self.reason = reason
        super().__init__(
            f"[pylva] {provider} request is not supported by the controlled provider "
            f"surface (reason={reason})"
        )

File: packages/sdk-ts/src/errors/strict_provider.ts
export class PylvaStrictProviderError extends TypeError {
  readonly code = 'strict_provider_unsupported' as const;

  constructor(
    readonly provider: 'openai' | 'anthropic' | 'vercel-ai',
    readonly reason: string,
  ) {
    super(`[pylva] strict ${provider} call refused: ${reason}`);
    this.name = 'PylvaStrictProviderError';
  }
}
Object.defineProperty(PylvaStrictProviderError, 'name', { value: 'PylvaStrictProviderError' });
